### ISSUE TITLE
Invalid Bidirectional Association Mapping with Mapped Superclass in Class Hierarchy

### ISSUE THREAD

**Issue by bobdercole** [doctrine/doctrine2#9516]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.11.1

#### Summary

I receive a mapping exception after defining an inverse-side association on an abstract entity. A mapped superclass extends the abstract entity. A concrete entity extends the mapped superclass.

#### Current behavior

ClassMetadataFactory is currently throwing an exception if there is an inverse-side association on a mapped superclass.

https://github.com/doctrine/orm/blob/e8e61cbbd56566c27bc5f766ac017d7921eb5577/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L381

This is correct, as stated in the [documentation](https://www.doctrine-project.org/projects/doctrine-orm/en/2.11/reference/inheritance-mapping.html#mapped-superclasses). However, I believe this limitation should be imposed only if the association is on the mapped superclass itself.

#### How to reproduce

- Define an abstract entity (A).
- Define a mapped superclass (B) extends A.
- Define a concrete entity (C) extends B.
- Define another entity (D) that has a many-to-one relationship with A.

I will follow-up with a failing test case.

#### Expected behavior

I think the above mapping should be valid since the association is not defined directly on the mapped superclass.

**Comment by bobdercole** [comment]
Here is my failing test case: #9517.

My proposed fix is to add a condition to [ClassMetadataFactory](https://github.com/doctrine/orm/blob/e8e61cbbd56566c27bc5f766ac017d7921eb5577/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L381) to only throw the exception if the association is directly on the mapped superclass. Maybe something like this?

```php
if ($mapping['type'] & ClassMetadata::TO_MANY && ! $mapping['isOwningSide'] && $parentClass->name === $mapping['sourceEntity']) {
    throw MappingException::illegalToManyAssociationOnMappedSuperclass($parentClass->name, $field);
}
```

**Comment by mpdude** [comment]
Related to #8415 (as per #9517).

**Comment by mpdude** [comment]
Please check if this is already fixed in the current 2.14.x branch

**Comment by mpdude** [comment]
You might also want to subscribe to #10398

**Comment by bobdercole** [comment]
> Please check if this is already fixed in the current 2.14.x branch

Yes! It works now. Thank you for all the work you've done to resolve it!

I suppose we should keep this open until the regression test PR is merged.

**Comment by mpdude** [comment]
@greg0ire maybe a low hanging fruit:

Merge #10453 and then close the issues and PRs linked from it
